Feat/keyframe communication

victor-amblard opened the pull request:
When a new keyframe is detected, it is added to a buffer that stores all the keyframes that have not been sent yet.
Then if some criterion is verified (for now there is none), keyframes are sent using KeyframePacket message format.
On ROS's side, topics `/rtabmap/kfListener_{id robot}` are used as publisher/subscribers.

What remains to be done: 

- Only send keyframes that make sense with regard to some geometric criteria
- Use range measurements read from messages to trigger keyframe transmission

Note that all the data structures are in `Memory.cpp` (the long-term memory) which we access through rtabmap's attribute `memory_`.

Commit:
Moved communication functions from Odometry to Memory

Commit:
Moved communication functions from Odometry to Memory

Commit:
Updated CI for current branch

Commit:
Updated CI for current branch

Commit:
Removed build dir

Commit:
Removed build dir

Commit:
Added nbRobots + robot's ID to parameters list

Commit:
Added nbRobots + robot's ID to parameters list

Commit:
Removed all previous class members from Rtabmap + proper variable initialization

Commit:
Removed all previous class members from Rtabmap + proper variable initialization

victor-amblard committed:
Merge branch 'master' into feat/keyframe_communication

alanpapalia reviewed (commented):
1. Are we expecting to just hardcode the number of robots in the swarm?
2. What is curLocalKFId intended to count?
3. For the functions added in Rtabmap.h  would it be possible to add brief comments to explain what the functions are actually doing

victor-amblard commented:
1. The number of robots is not really hardcoded (in the sense that you don't have to recompile the code to change the number of robots), it's something that you add in the launch file. If you have a look at the folder example/four_robots.launch in rtabmap_ros, master branch, you can see that I just add a line with nbRobots and the robot's id. It seemed like the best way of doing it for me but I'm open to suggestions! 

2. curLocalKFId counts the number of keyframes that have been detected by the robot so far. The reason why I'm using this is to avoid directly storing the keyframes' descriptors N times where N is the number of robots and rather use one structure that 
stores the keyframes with their id `allLocalDescriptors` and identify the keyframes in the transmission queues as their id's.

3. For sure! I'll push comments (cf commit `Added comments and removed unused statement`). Actually there's almost no new code in Rtabmap.h so I guess you meant Rtabmap.cpp or Memory.h

Commit:
Added comments and removed unused statement

Commit:
Fixed conflicts on local repo

Commit:
Fixed errors when fixing conflicts

Commit:
Fix segfault keyframe

victor-amblard commented:
Added fix for a segfault occurring while transmitting keyframes.
It's not completely clear to me what the underlying process is yet, but it looks like that variables in rtabmap are being erased at each iteration. So `bufferCommunicationKF` which is the variable in which keyframes that need to be transmitted to other robots are stored, was erased => segfault.
I moved the variable and the associated functions to `Memory.*` to avoid segfaults and it seems to work...

Commit:
Added wrapper for cleanTransmittedKF (needed by rtabmap_ros)
